examples/rabbit_fox_hawk.jl

```diff
@@ -0,0 +1,428 @@
+        if heightmap[i, j] < mountain_level
+            air_walkmap[i, j, (heightmap[i, j]+1):mountain_level] .= true
+        end
+    end
+
+    ## Generate the RNG for the model
+    rng = MersenneTwister(seed)
+
+    ## Note that the dimensions of the space do not have to correspond to the dimensions
+    ## of the pathfinder. Discretisation is handled by the pathfinding methods
+    space = ContinuousSpace((100., 100., 100.); periodic = false)
+
+    ## Generate an array of random numbers, and threshold it by the probability of grass growing
+    ## at that location. Although this causes grass to grow below `water_level`, it is
+    ## effectively ignored by `land_walkmap`
+    grass = BitArray(
+        rand(rng, dims[1:2]...) .< ((grass_level .- heightmap) ./ (grass_level - water_level)),
+    )
+    properties = (
+        ## The pathfinder for rabbits and foxes
+        landfinder = AStar(space, land_walkmap),
+        ## The pathfinder for hawks
+        airfinder = AStar(space, air_walkmap; cost_metric = MaxDistance{3}()),
+        Δe_grass = Δe_grass,
+        Δe_rabbit = Δe_rabbit,
+        rabbit_repr = rabbit_repr,
+        fox_repr = fox_repr,
+        hawk_repr = hawk_repr,
+        rabbit_vision = rabbit_vision,
+        fox_vision = fox_vision,
+        hawk_vision = hawk_vision,
+        rabbit_speed = rabbit_speed,
+        fox_speed = fox_speed,
+        hawk_speed = hawk_speed,
+        heightmap = heightmap,
+        grass = grass,
+        regrowth_chance = regrowth_chance,
+        water_level = water_level,
+        grass_level = grass_level,
+        dt = dt,
+    )
+
+    model = ABM(Animal, space; rng, properties)
+
+    ## spawn each animal at a random walkable position according to its pathfinder
+    for _ in 1:n_rabbits
+        add_agent_pos!(
+            Rabbit(
+                nextid(model), ## Using `nextid` prevents us from having to manually keep track
+                               ## of animal IDs
+                random_walkable(model, model.landfinder),
+                rand(model.rng, Δe_grass:2Δe_grass),
+            ),
+            model,
+        )
+    end
+    for _ in 1:n_foxes
+        add_agent_pos!(
+            Fox(
+                nextid(model),
+                random_walkable(model, model.landfinder),
+                rand(model.rng, Δe_rabbit:2Δe_rabbit),
+            ),
+            model,
+        )
+    end
+    for _ in 1:n_hawks
+        add_agent_pos!(
+            Hawk(
+                nextid(model),
+                random_walkable(model, model.airfinder),
+                rand(model.rng, Δe_rabbit:2Δe_rabbit),
+            ),
+            model,
+        )
+    end
+
+    return model
+end
+
+# We create a utility function that takes the 3D position of an agent, and returns the
+# corresponding 2D position on the heightmap
+to_heightmap_index(pos, model) =
+    floor.(Int, pos[1:2] ./ size(model.space)[1:2] .* size(model.heightmap)) .+ 1
```

Comment:
I just pushed to `master` branch an update to `get_spatial_index` that will make this not necessary.